src/rust/ide/view/graph-editor/src/component/node/port.rs

```diff
@@ -233,54 +284,62 @@ impl Manager {
                     let contains_root = span.index.value == 0;
                     let skip          = node.kind.is_empty() || contains_root;
                     if !skip {
-                        let logger   = Logger::sub(&self.logger,"port");
+                        let logger   = Logger::sub(&model.logger,"port");
                         let port     = component::ShapeView::<shape::Shape>::new(&logger,self.scene());
-                        let type_map = &self.type_color_map;
+                        let type_map = &model.type_color_map;
 
                         let unit        = 7.224_609_4;
                         let width       = unit * span.size.value as f32;
                         let width2      = width + 8.0;
                         let node_height = 28.0;
                         let height      = 18.0;
+                        let size        = Vector2::new(width2,height);
                         port.shape.sprite.size.set(Vector2::new(width2,node_height));
                         let x = width/2.0 + unit * span.index.value as f32;
                         port.mod_position(|t| t.x = x);
-                        self.add_child(&port);
-
-                        let hover   = &port.shape.hover;
-                        let crumbs  = node.crumbs.clone();
-                        let ast_id  = get_id_for_crumbs(&expression.input_span_tree,&crumbs);
+                        model.ports_group.add_child(&port);
 
                         // FIXME : StyleWatch is unsuitable here, as it was designed as an internal tool for shape system (#795)
-                        let styles             = StyleWatch::new(&self.app.display.scene().style_sheet);
+                        let styles             = StyleWatch::new(&model.app.display.scene().style_sheet);
                         let missing_type_color = styles.get_color(theme::vars::graph_editor::edge::_type::missing::color);
 
+                        let crumbs = node.crumbs.clone();
+                        let ast_id = get_id_for_crumbs(&expression.input_span_tree,&crumbs);
+                        let color  = ast_id.and_then(|id|type_map.type_color(id,styles.clone_ref()));
+                        let color  = color.unwrap_or(missing_type_color);
+
+                        let highlight = cursor::Style::new_highlight(&port,size,Some(color));
+
                         frp::new_network! { port_network
```

Comment:
As I understand, when someone will capture "model" in some of nodes in this network we risk the leak. And this is quite obscure property of this piece of code. I would refactor the creating "port_network" into a separate function, which does not take model of Port Manager, and add explicit explanation that it must not do.

Reply:
Lets talk tomorrow about it

Reply:
Planned for creating strict-owned network (without CloneRef).